launcher/net/PasteUpload.cpp

```diff
@@ -19,26 +69,73 @@ PasteUpload::~PasteUpload()
 
-    QHttpMultiPart *multiPart = new QHttpMultiPart{QHttpMultiPart::FormDataType};
+    switch (m_pasteType) {
+    case NullPointer: {
+        QHttpMultiPart *multiPart =
+          new QHttpMultiPart{QHttpMultiPart::FormDataType};
 
-    QHttpPart filePart;
-    filePart.setBody(m_text);
-    filePart.setHeader(QNetworkRequest::ContentTypeHeader, "text/plain");
-    filePart.setHeader(QNetworkRequest::ContentDispositionHeader, "form-data; name=\"file\"; filename=\"log.txt\"");
+        QHttpPart filePart;
+        filePart.setBody(m_text);
+        filePart.setHeader(QNetworkRequest::ContentTypeHeader, "text/plain");
+        filePart.setHeader(QNetworkRequest::ContentDispositionHeader,
+                         "form-data; name=\"file\"; filename=\"log.txt\"");
+        multiPart->append(filePart);
 
-    multiPart->append(filePart);
+        rep = APPLICATION->network()->post(request, multiPart);
+        multiPart->setParent(rep);
 
-    QNetworkReply *rep = APPLICATION->network()->post(request, multiPart);
-    multiPart->setParent(rep);
+        break;
+    }
+    case Hastebin: {
+        request.setHeader(QNetworkRequest::UserAgentHeader, BuildConfig.USER_AGENT_UNCACHED);
+        rep = APPLICATION->network()->post(request, m_text);
+        break;
+    }
+    case Mclogs: {
+        QUrlQuery postData;
+        postData.addQueryItem("content", m_text);
+        request.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
+        rep = APPLICATION->network()->post(request, postData.toString().toUtf8());
+        break;
+    }
+    case PasteGG: {
+        QJsonObject obj;
+        QJsonDocument doc;
+        request.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
 
-    m_reply = std::shared_ptr<QNetworkReply>(rep);
-    setStatus(tr("Uploading to %1").arg(m_uploadUrl));
+        obj.insert("expires", QDateTime::currentDateTimeUtc().addDays(100).toString(Qt::DateFormat::ISODate));
+
+        QJsonArray files;
+        QJsonObject logFileInfo;
+        QJsonObject logFileContentInfo;
+        logFileContentInfo.insert("format", "text");
+        logFileContentInfo.insert("value", QString::fromUtf8(m_text));
+        logFileInfo.insert("name", "log.txt");
+        logFileInfo.insert("content", logFileContentInfo);
+        files.append(logFileInfo);
+
+        obj.insert("files", files);
+
+        doc.setObject(obj);
+        rep = APPLICATION->network()->post(request, doc.toJson());
+        break;
+    }
+    }
 
     connect(rep, &QNetworkReply::uploadProgress, this, &Task::setProgress);
-    connect(rep, SIGNAL(error(QNetworkReply::NetworkError)), this, SLOT(downloadError(QNetworkReply::NetworkError)));
-    connect(rep, SIGNAL(finished()), this, SLOT(downloadFinished()));
+    connect(rep, &QNetworkReply::finished, this, &PasteUpload::downloadFinished);
+    // This function call would be a lot shorter if we were using the latest Qt
```

Comment:
What Qt version is this targeting? Most of the codebase uses the `SIGNAL` and `SLOTG` macros

Reply:
As far as I know, those macros connect the signals and slots by name, at runtime, not at compile-time. So even though we use those macros throughout the codebase, it's safer and more performant to define it like this, whenever we can.

Reply:
Oh. I just read up again. Was looking at the wrong documentation.

Reply:
> What Qt version is this targeting? Most of the codebase uses the `SIGNAL` and `SLOTG` macros

Pointer to member function is the modern way of doing it and has some advantages over the signal and slot macros. It's supported by the Qt versions we support.